src/webgpu/shader/execution/shader_io/fragment_builtins.spec.ts

```diff
@@ -869,7 +941,124 @@ g.test('inputs,sample_index')
         sampleCount,
         actual,
         expected,
-        maxFractionalDiff: 0.00001,
+        maxDiffULPsForFloatFormat: 1,
+      })
+    );
+  });
+
+g.test('inputs,front_facing')
+  .desc(
+    `
+    Test fragment shader builtin(front_facing) values.
+
+    Draws a quad from 2 triangles that entirely cover clip space. (see diagram below in code)
+    One triangle is clockwise, the other is counter clockwise. The triangles
+    bisect pixels so that different samples are covered by each triangle so that some
+    samples should get different values for front_facing for the same fragment.
+  `
+  )
+  .params(u =>
+    u //
+      .combine('nearFar', [[0, 1] as const, [0.25, 0.75] as const] as const)
+      .combine('sampleCount', [1, 4] as const)
+      .combine('frontFace', ['cw', 'ccw'] as const)
+      .combine('interpolation', [
+        { type: 'perspective', sampling: 'center' },
+        { type: 'perspective', sampling: 'centroid' },
+        { type: 'perspective', sampling: 'sample' },
+        { type: 'linear', sampling: 'center' },
+        { type: 'linear', sampling: 'centroid' },
+        { type: 'linear', sampling: 'sample' },
+        { type: 'flat' },
+      ] as const)
+  )
+  .beforeAllSubcases(t => {
+    const {
+      interpolation: { type, sampling },
+    } = t.params;
+    t.skipIfInterpolationTypeOrSamplingNotSupported({ type, sampling });
+  })
+  .fn(async t => {
+    const {
+      nearFar,
+      sampleCount,
+      frontFace,
+      interpolation: { type, sampling },
+    } = t.params;
+    //
+    // We're drawing 2 triangles starting at y = -2 to y = +2
+    //
+    //  -1   0   1
+    //   +===+===+  2
+    //   |   |  /|
+    //   +---+---+  1  <---
+    //   |   |/  |       |
+    //   +---+---+  0    | viewport
+    //   |  /|   |       |
+    //   +---+---+ -1  <---
+    //   |/  |   |
+    //   +===+===+ -2
+
+    // prettier-ignore
+    const clipSpacePoints = [
+      // cw
```

Comment:
Does the "cw" and "ccw" comment also need to flip here?

Reply:
done